runcoverage.sh

```diff
@@ -23,40 +23,67 @@ then
   fi
 fi
 
-$NUGET install -Verbosity quiet -OutputDirectory packages -Version 4.6.519 OpenCover
+$NUGET install -Verbosity quiet -OutputDirectory packages -Version 2017.1.20170613.162720 JetBrains.dotCover.CommandLineTools
 $NUGET install -Verbosity quiet -OutputDirectory packages -Version 2.4.5.0 ReportGenerator
-OPENCOVER=$PWD/packages/OpenCover.4.6.519/tools/OpenCover.Console.exe
+
+DOTCOVER=$PWD/packages/JetBrains.dotCover.CommandLineTools.2017.1.20170613.162720/tools/dotCover.exe
 REPORTGENERATOR=$PWD/packages/ReportGenerator.2.4.5.0/tools/ReportGenerator.exe
 FIND=/usr/bin/find
 
-rm -rf coverage
-mkdir coverage
+if [ ! -d coverage ]
+then
+ mkdir coverage
+fi
 
 while read csproj
 do
   testdir=$(dirname $csproj)
   api=$(basename $(realpath $testdir/..))
+
+  [ ! -d "$testdir/bin/Release/net452" ] && [ ! -d "$testdir/bin/Release/netcoreapp1.0" ] && continue
+
   echo "Running coverage for $api"
-  # OpenCover is picky about how it finds the excluded files. There may be a better approach than this,
```

Comment:
Hmm. Good to see the attribute. It feels slightly awkward to add that to the generators for a few reasons:

- There are three to add them to - protobuf, gRPC, toolkit - which means they'll be rolled out at different rates
- The toolkit code is meant to look like "the code you'd write by hand if you had to" as far as possible; this will definitely be uglier
- I'm still not sure how we handle the protobuf code like this:

        private static readonly pb::MessageParser<Policy> _parser = new pb::MessageParser<Policy>(() => new Policy());

 There's one of those per proto message, and the lambda expression doesn't get covered unless the parser is used. We can't put the attribute on the field declaration, and I don't *think* you could apply it to the lambda either. Will have to investigate further.

All that said, it seems reasonable that other users may well want to do code coverage involving proto and gRPC generated code too. It's just such a shame when filtering by source code does *exactly* what we want :(